<a href="https://colab.research.google.com/github/nanytahm/google-colab/blob/main/datosInfolimpieza.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

                 **Nombre: NANCY HUAMAN MAMANI**
#                   LISTA PARA INFO-LIMPIEZA

##Cargar librerias a usar



In [1]:
import pandas as pd
import numpy as np

##cargar datos de formato excel a un dataframe de pandas

###nombramos df a nuestra nueva tabla


In [2]:
df=pd.read_excel("/content/ALUMNO_INFORM_2022_2.xlsx")
df.head()

Alumno                   NombreAlumno codigocurso  \
0  131612       MONTAÑEZ-CHOQUE-WILLIANS    IF612BIN   
1  174944  CONDORI-MOZO-WESLEY JUANPEDRO    IF063AIN   
2  183885        CHILE-QUIROGA-HILDEMARO    IF553AIN   
3  221945   ALVAREZ-CATUNTA-ANGEL ISMAEL    FI902AIN   
4  171879           QUISPE-MAMANI-THALIA    IF610AIN   

                                    Nombrecurso  Docente  \
0        FUNDAMENTOS Y DISEÑO DE BASES DE DATOS  14768.0   
1                                       QUECHUA      NaN   
2                          LENGUAJE ENSAMBLADOR  13733.0   
3                                      FÍSICA I  10067.0   
4  ANALISIS Y DISEÑO DE SISTEMAS DE INFORMACIÓN  14664.0   

                  NombreDocente  
0        MEDINA-MIRANDA-KARELIA  
1                           NaN  
2       PALOMINO-OLIVERA-EMILIO  
3  BARRIENTOS-ECHEGARAY-HONORIO  
4     CANDIA-OVIEDO-DENNIS IVAN

In [3]:
#ver cantidad de datos que tenemos
df.shape

(3649, 6)

##cantidad de nulos que hay en cada columna

In [4]:
df.isna().sum()

Alumno             0
NombreAlumno       0
codigocurso        0
Nombrecurso        0
Docente          886
NombreDocente    886
dtype: int64

#Observando la consulta anterior, inferimos que hay que eliminar 886 filas

In [76]:
ndf=df.dropna()
ndf

Alumno                       NombreAlumno codigocurso  \
0     131612           MONTAÑEZ-CHOQUE-WILLIANS    IF612BIN   
2     183885            CHILE-QUIROGA-HILDEMARO    IF553AIN   
3     221945       ALVAREZ-CATUNTA-ANGEL ISMAEL    FI902AIN   
4     171879               QUISPE-MAMANI-THALIA    IF610AIN   
5     193001  GIFONE-VILLASANTE-EDUARDO JUAREIS    IF484AIN   
...      ...                                ...         ...   
3643  204798       HUARACALLO-ARENAS-LINO ZEYNT    FI370AIN   
3644  194518              HUAMAN-JAIMES-NICANOR    IF651AIN   
3646  220551       QUESADA-MAMANI-RODRIGO ANDRE    IF468BIN   
3647  193832    CORTEZ-CCAHUANTICO-PAOLA ANDREA    IF614BIN   
3648  184689              PILLCO-FLORES-LISBETH    IF459BIN   

                                       Nombrecurso  Docente  \
0           FUNDAMENTOS Y DISEÑO DE BASES DE DATOS  14768.0   
2                             LENGUAJE ENSAMBLADOR  13733.0   
3                                         FÍSICA I  10067.0   
4     ANALISIS Y DISEÑO DE SISTEMAS DE INFORMACIÓN  14664.0   
5                      EMPRENDIMIENTO E INNOVACION  14691.0   
...                                            ...      ...   
3643                               FISICA APLICADA  10208.0   
3644                       INTELIGENCIA ARTIFICIAL  81325.0   
3646                   FUNDAMENTOS DE PROGRAMACIÓN  11849.0   
3647                      INGENIERIA DE SOFTWARE I  16571.0   
3648                        COMPUTACION GRAFICA II  14995.0   

                       NombreDocente  
0             MEDINA-MIRANDA-KARELIA  
2            PALOMINO-OLIVERA-EMILIO  
3       BARRIENTOS-ECHEGARAY-HONORIO  
4          CANDIA-OVIEDO-DENNIS IVAN  
5        CHAVEZ-CENTENO-JAVIER DAVID  
...                              ...  
3643   DE LA TORRE-CRUZ-JOSE OLINTHO  
3644           ZUÑIGA-ROJAS-GABRIELA  
3646      IBARRA-ZANBRANO-WALDO ELIO  
3647  DUEÑAS-DE LA CRUZ-HENRY SAMUEL  
3648          VILLAFUERTE-SERNA-RONY  

[2763 rows x 6 columns]

In [6]:
#se reasignó nuestros datos a una nueva tabla "ndf" 
#observamos la cantidad de filas y columnas que nos quedan
ndf.shape

(2763, 6)

##Actualizar ndf con estudiantes y docentes que son solo de informatica

In [7]:
ndf2=ndf[ndf.apply(lambda row: row['codigocurso'].startswith('IF'),axis=1)]

In [8]:
ndf2.sort_values(by='codigocurso')

Alumno                        NombreAlumno codigocurso  \
1176  210410           PACCO-MAMANI-MARLOM DIEGO    IF060AIN   
3008  211860                MUÑOZ-CENTENO-MILDER    IF060AIN   
919   211356      PALACIOS-DURAND-ORIOL FERNANDO    IF060AIN   
1713  204807            TINOCO-CCOTO-LUIS MANUEL    IF060AIN   
2112  133960         CHAMPI-CHAMPI-ABRAHAN ELIAS    IF060AIN   
...      ...                                 ...         ...   
1692  221446                   DIAZ-MISME-PAMELA    IF902BIN   
1408  220213  PUMACCAHUA-HUALLPA-PATRICK MICHAEL    IF902BIN   
1245  215785            FERNANDEZ-PUMA-SEBASTIAN    IF902BIN   
39    221448         GIL-FIGUEROA-HEIDAN TORIBIO    IF902BIN   
902   164563          CARPIO-HERMOZA-HAIDER ALEX    IF902BIN   

                                          Nombrecurso  Docente  \
1176                                           MÚSICA  81145.0   
3008                                           MÚSICA  81145.0   
919                                            MÚSICA  81145.0   
1713                                           MÚSICA  81145.0   
2112                                           MÚSICA  81145.0   
...                                               ...      ...   
1692  TECNOLOGIAS DE LA INFORMACIÓN Y LA COMUNICACIÓN  81145.0   
1408  TECNOLOGIAS DE LA INFORMACIÓN Y LA COMUNICACIÓN  81145.0   
1245  TECNOLOGIAS DE LA INFORMACIÓN Y LA COMUNICACIÓN  81145.0   
39    TECNOLOGIAS DE LA INFORMACIÓN Y LA COMUNICACIÓN  81145.0   
902   TECNOLOGIAS DE LA INFORMACIÓN Y LA COMUNICACIÓN  81145.0   

           NombreDocente  
1176  CHULLO-LLAVE-BORIS  
3008  CHULLO-LLAVE-BORIS  
919   CHULLO-LLAVE-BORIS  
1713  CHULLO-LLAVE-BORIS  
2112  CHULLO-LLAVE-BORIS  
...                  ...  
1692  CHULLO-LLAVE-BORIS  
1408  CHULLO-LLAVE-BORIS  
1245  CHULLO-LLAVE-BORIS  
39    CHULLO-LLAVE-BORIS  
902   CHULLO-LLAVE-BORIS  

[2212 rows x 6 columns]

In [9]:
#nuestra tabla actualizada con solo estudiantes y docentes de INFORMATICA tiene 2212 filas x 6 columnas
ndf2.shape

(2212, 6)



*   SECCION PARA VER CANTIDAD DE ALUMNOS, CURSOS Y DOCENTES DE INFORMATICA



In [10]:
#CANTIDAD DE CURSOS: 70
pd.DataFrame(ndf2[['codigocurso','Nombrecurso']].value_counts()).sort_values(by='Nombrecurso')

0
codigocurso Nombrecurso                                        
IF450BIN    ABSTRACCIÓN DE DATOS Y OBJETOS                   13
IF450AIN    ABSTRACCIÓN DE DATOS Y OBJETOS                   39
IF480BIN    ADMINISTRACIÓN DE TECNOLOGIAS DE INFORMACIÓN      9
IF480AIN    ADMINISTRACIÓN DE TECNOLOGIAS DE INFORMACIÓN     40
IF456BIN    ALGORITMOS AVANZADOS                             11
...                                                          ..
IF064AIN    TALLER  DE DEBATE                                37
IF902AIN    TECNOLOGIAS DE LA INFORMACIÓN Y LA COMUNICACIÓN  22
IF902BIN    TECNOLOGIAS DE LA INFORMACIÓN Y LA COMUNICACIÓN  21
IF454AIN    TEORIA DE LA COMPUTACION                         42
IF657AIN    VISION COMPUTACIONAL                             41

[70 rows x 1 columns]

In [11]:
##NUMERO TOTAL DE ESTUDIANTES:485
pd.DataFrame(ndf2[['Alumno','NombreAlumno']].value_counts()).sort_values(by='NombreAlumno')

0
Alumno NombreAlumno                         
151450 ACHAHUANCO-ACHAHUI-EURIDICE INGRID  5
204792 ACHAHUANCO-VALENZA-ANDREE           5
201228 ACHAHUI-CRUZ-MILTON AMED            6
182893 AGUILAR-PORCEL-JAZMIN               4
204793 AGUILAR-SANCHEZ-NIK ANTONI          5
...                                       ..
161534 ZEGARRA-ROJAS-JORGE ENRIQUE         4
194530 ZEVALLOS-VIDAL-NYCOLL TATIANA       7
204322 ZULOAGA-CCOPA-NILSON LEONEL         2
182942 ZUNIGA-AUCCAHUAQUI-JUAN MANUEL      5
192190 ZUÑIGA-SARA-CARLOS DANIEL           3

[485 rows x 1 columns]

In [36]:
##NÚMERO TOTAL DE DOCENTES:41
pd.DataFrame(ndf2[['Docente','NombreDocente']].value_counts()).sort_values(by='NombreDocente').head(10)

0
Docente NombreDocente                     
11203.0 ACURIO-USCA-NILA ZONIA          48
14623.0 AGUIRRE-CARBAJAL-DORIS SABINA   61
14769.0 ALZAMORA-PAREDES-ROBERT         70
12459.0 BACA-CARDENAS-LINO AQUILES      20
14664.0 CANDIA-OVIEDO-DENNIS IVAN       77
13226.0 CARBAJAL-LUNA-JULIO CESAR       34
13428.0 CARRASCO-POBLETE-EDWIN          44
14691.0 CHAVEZ-CENTENO-JAVIER DAVID     30
81165.0 CHOQUE-SOTO-VANESSA MARIBEL     62
81145.0 CHULLO-LLAVE-BORIS             112

In [38]:
pd.DataFrame(ndf2[['Docente','NombreDocente']].value_counts()).size


41

# funciones



In [19]:
tDocente=ndf2['Docente'].unique()
tAlumno=ndf2['Alumno'].unique()

In [ ]:
# iniciamos un diccionario, index con codigo docente y values 0 
# que se modificará conforme se agreguen alumnos a la lista del docente
nro=np.zeros(41)
cantAlumnDocente=dict(zip(tDocente,nro))
cantAlumnDocente

In [21]:
#inicializacion de listas que usaremos:
#lista estudiantes
listEstudianteA=[]
#diccionario donde se iran agregando el alumno y docente para la lista de la infolimpieza
listAlumnoDocente={}

In [22]:
#funcion para saber si el codigo del alumno que estamos viendo
#ya a sido asignado con un docente

def tienedocente(alumno):
  if alumno in listEstudianteA:
    return True
  else:
    return False

In [23]:
##funcion para asignarle docente a cada alumno

#teniendo la cantidad de estudiantes y de alumnos, 
#calculamos que debe asignarse a lo mas 15 alumnos a cada docente

def asignardocente(a,d):
  if (tienedocente(a)==True):
    return listAlumnoDocente[a] 
  elif (cantAlumnDocente[d] < 15):
      cant=cantAlumnDocente[d]
      listEstudianteA.append(a)
      cantAlumnDocente[d] = cant + 1
      listAlumnoDocente[a] = d
      return d
  else:
      return np.NaN



In [ ]:
#se crea una columna para cada alumno en la que
#se tenga el codigo del docente al que irá pertnecer a su lista
 
ndf2.apply(lambda x: asignardocente(x[0],x[4]), axis=1,  result_type='expand')



*   SECCION DE RESULTADOS





a. Lista de Docentes de Informática



In [31]:
listadocentes=pd.DataFrame(ndf2[['Docente','NombreDocente']].value_counts()).sort_values(by='NombreDocente')


In [56]:
#lista de docentes, ordenado por apellido
listadocentes


0
Docente NombreDocente                           
11203.0 ACURIO-USCA-NILA ZONIA                48
14623.0 AGUIRRE-CARBAJAL-DORIS SABINA         61
14769.0 ALZAMORA-PAREDES-ROBERT               70
12459.0 BACA-CARDENAS-LINO AQUILES            20
14664.0 CANDIA-OVIEDO-DENNIS IVAN             77
13226.0 CARBAJAL-LUNA-JULIO CESAR             34
13428.0 CARRASCO-POBLETE-EDWIN                44
14691.0 CHAVEZ-CENTENO-JAVIER DAVID           30
81165.0 CHOQUE-SOTO-VANESSA MARIBEL           62
81145.0 CHULLO-LLAVE-BORIS                   112
18428.0 CUTIPA-ARAPA-EFRAINA GLADYS           37
16570.0 DUEÑAS-BUSTINZA-DARIO FRANCISCO       46
16571.0 DUEÑAS-DE LA CRUZ-HENRY SAMUEL        52
81809.0 DUEÑAS-JIMENEZ-RAY                    36
10234.0 ENCISO-RODAS-LAURO                    44
13994.0 FLORES-PACHECO-LINO PRISCILIANO       57
81629.0 GAMARRA-SALAS-JISBAJ                  17
12004.0 GAMARRA-SALDIVAR-ENRIQUE              20
80246.0 HUILLCA-HUALLPARIMACHI-RAUL           16
11849.0 IBARRA-ZANBRANO-WALDO ELIO            16
17093.0 IRPANOCCA-CUSIMAYTA-MARITZA           71
14768.0 MEDINA-MIRANDA-KARELIA                39
12596.0 MEDRANO-VALENCIA-IVAN CESAR           71
81773.0 MONTOYA-CUBAS-CARLOS FERNANDO         98
15938.0 MONZON-CONDORI-LUIS ALVARO            61
14767.0 ORMEÑO-AYALA-YESHICA ISELA            56
16563.0 PALMA-TTITO-LUIS BELTRAN              46
13733.0 PALOMINO-OLIVERA-EMILIO               87
14627.0 PEÑALOZA-FIGUEROA-MANUEL AURELIO      41
14002.0 PILLCO-QUISPE-JOSE MAURO              60
81630.0 QUINTANILLA-PORTUGAL-ROXANA LISETTE   30
80414.0 QUISPE-ONOFRE-CARLOS RAMON            22
13225.0 ROZAS-HUACHO-JAVIER ARTURO            74
81311.0 SEGUNDO-CARPIO-LISETH URPY           106
14389.0 TICONA-PARI-GUZMAN                    33
81474.0 UGARTE-ROJAS-HECTOR EDUARDO           90
81146.0 VERA-OLIVERA-HARLEY                   43
14995.0 VILLAFUERTE-SERNA-RONY                80
81262.0 VILLALBA-VILLALBA-TANY                50
80629.0 ZAMALLOA-PARO-WILLIAN                 63
81325.0 ZUÑIGA-ROJAS-GABRIELA                 92


b.  Lista de Alumnos

In [57]:
#lista alumnos, ordenado por apellido
listaalumnos=pd.DataFrame(ndf2[['Alumno','NombreAlumno']].value_counts()).sort_values(by='NombreAlumno')

In [59]:
listaalumnos

0
Alumno NombreAlumno                         
151450 ACHAHUANCO-ACHAHUI-EURIDICE INGRID  5
204792 ACHAHUANCO-VALENZA-ANDREE           5
201228 ACHAHUI-CRUZ-MILTON AMED            6
182893 AGUILAR-PORCEL-JAZMIN               4
204793 AGUILAR-SANCHEZ-NIK ANTONI          5
...                                       ..
161534 ZEGARRA-ROJAS-JORGE ENRIQUE         4
194530 ZEVALLOS-VIDAL-NYCOLL TATIANA       7
204322 ZULOAGA-CCOPA-NILSON LEONEL         2
182942 ZUNIGA-AUCCAHUAQUI-JUAN MANUEL      5
192190 ZUÑIGA-SARA-CARLOS DANIEL           3

[485 rows x 1 columns]

c. Lista de docentes con sus alumnos asignados

In [60]:
dfn=pd.DataFrame(list(listAlumnoDocente.items()), columns=['Alumno', 'Docente'])
dfn.head(5)

Alumno  Docente
0  131612  14768.0
1  183885  13733.0
2  171879  14664.0
3  193001  14691.0
4  215270  18428.0

In [75]:
#agrupamos para ver docente con su lista de alumnos
pd.DataFrame(dfn.groupby(["Docente", "Alumno"])['Alumno'].size())


Alumno
Docente Alumno        
10234.0 40986        1
        140997       1
        141158       1
        150404       1
        155637       1
...                ...
81809.0 182909       1
        182910       1
        182925       1
        182927       1
        182936       1

[456 rows x 1 columns]

In [74]:
#prueba para ver lista de cada docente por codigodocente
dfn[dfn["Docente"]==10234]

Alumno  Docente
18   140997  10234.0
68   204318  10234.0
88   150404  10234.0
116  184656  10234.0
180  204801  10234.0
281  141158  10234.0
325  155637  10234.0
383  200340  10234.0
434  184193  10234.0
444  183067  10234.0
453  184197  10234.0
454   40986  10234.0